### PR TITLE
Fault tolerant mechanisms

### LINKED ISSUE #1: implement fault tolerant mechanisms
Specifically:

- When no subscribers read a message, the publisher should archive it, in order for the pubsub to read it when restarting
- When a connection error occurs (for ex. disconnection), the listener should resubscribe automatically

### PR DESCRIPTION
**Pull request by J0sueTM**
closes #1 

## Summary by CodeRabbit

- **New Features**
  - Added message archiving and unarchiving functionalities.
  - Introduced worker subscription to channels with reconnection capabilities.

- **Enhancements**
  - Improved message publishing by considering consumer count.
  - Grouped handlers by channel for better management.

- **Refactor**
  - Consolidated parameters into an `options` parameter for pushing and popping messages in queues.

- **Tests**
  - Added tests for new archiving/unarchiving and multi pub/sub features.
  - Refactored existing tests for improved coverage and clarity.

### CODE REVIEW

File: src/com/moclojer/rq/pubsub.clj [hunk 1]
```diff
@@ -1,43 +1,168 @@
...
+        (recur (inc message-count)))
+      (log/debug "unarchived channel"
+                 {:channel channel
+                  :message-count message-count}))))
+
+(defn pack-workers-channels
+  [workers]
+  (map #(update % :channel (partial utils/pack-pattern :pubsub)) workers))
+
+(comment
```
Comment (avelino) [inline src/com/moclojer/rq/pubsub.clj:96]: move comment to tests pls

File: src/com/moclojer/rq/pubsub.clj [hunk 2]
```diff
@@ -1,43 +1,168 @@
...
+
+    (doseq [channel (map :channel workers)]
+      (unarquive-channel! client channel
+                          #(.onMessage listener
+                                       (utils/pack-pattern :pubsub channel) %)))
+
+    (future
+      (try
+        (.subscribe @client listener (into-array packed-channels))
+        (prn :hmmm packed-channels opts)
```
Comment (avelino) [inline src/com/moclojer/rq/pubsub.clj:131]: debug or log?

File: src/com/moclojer/rq/pubsub.clj [hunk 3]
```diff
@@ -1,43 +1,168 @@
...
+        (log/debug "subscribed workers to channels"
+                   {:channels packed-channels
+                    :options opts})
+        (catch JedisConnectionException e
+          (prn :e (.getMessage e))
```
Comment (avelino) [inline src/com/moclojer/rq/pubsub.clj:136]: remove debug!

File: src/com/moclojer/rq/pubsub.clj [hunk 4]
```diff
@@ -1,43 +1,168 @@
...
+          (log/warn "subscriber connection got killed. trying to reconnect..."
+                    {:channels packed-channels
+                     :exception e
+                     :ex-message (.getMessage e)})
+          (Thread/sleep reconnect-sleep))))
+
+    listener))
+
+(comment
```
Comment (avelino) [inline src/com/moclojer/rq/pubsub.clj:145]: Oh my good!
I get chills seeing debug _(comment)_ in the code

File: src/com/moclojer/rq/utils.clj [hunk 5]
```diff
@@ -2,8 +2,10 @@
 
 (defn- pattern->str
```
Comment (avelino) [inline src/com/moclojer/rq/utils.clj:3]: write test and docstring

**Review by avelino — CHANGES_REQUESTED** [review]

File: src/com/moclojer/rq/queue.clj [hunk 6]
```diff
@@ -7,13 +7,10 @@
...
-  - client: Redis client
-  - queue-name: Name of the queue
-  - message: Message to be pushed
-  - options: Optional parameters, including:
-    - direction: Direction to push the message (:l or :r)
-    - pattern: Pattern for the queue name"
+  Options:
+
+  - direction: Direction to push the message (:l or :r)
+  - pattern: Pattern for the queue name"
```
[the remark this answers is not in the document]
Comment (J0sueTM) [reply src/com/moclojer/rq/queue.clj:13]: @coderabbitai Options are passed directly, not in a map, as you though. There's nothing wrong with what you're seeing.

**Pull request merged**